### PR TITLE
feat(eslint-plugin): add no-unnecessary-generic-modifier rule

### LINKED ISSUE #598: Rule proposal: Prevent moot `unknown` constraint/default in generics
In TypeScript 3.5, [generic type parameters are implicitly constrained to `unknown` instead of `{}`](https://devblogs.microsoft.com/typescript/announcing-typescript-3-5/#breaking-changes). It would be nice if this rule could warn about moot usage of that.

**Expected Result**

I expected these uses to be reported as unnecessary:

```ts
function set<T extends unknown>(object: T): T {}

function set<T = unknown>(object: T): T {}

set<unknown>(object: {foo: true});
```

**Actual Result**

They are not.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.9.0` |
| `@typescript-eslint/parser`        | `1.9.0` |
| `TypeScript`                       | `3.5.1` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `8.16.0` |
| `npm`                              | `6.9.0` |

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

-   [x] Addresses an existing issue: fixes #598
-   [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
-   [x] Steps in [CONTRIBUTING.md](https://github.com/typescript-eslint/typescript-eslint/blob/main/CONTRIBUTING.md) were taken

## Overview

Adds a fixable `@typescript-eslint/no-redundant-generic-modifiers` rule that flags `extends unknown` and `= unknown` generic type parameters. The only allowed case is `= unknown` in a JSX syntax area.

Co-authored-by: @anthonyronda

Interestingly, by renaming a rule in the recommended set, this is a breaking change.

### CODE REVIEW

File: packages/eslint-plugin/src/rules/no-redundant-generic-modifiers.ts [hunk 1]
```diff
@@ -0,0 +1,91 @@
+import { AST_NODE_TYPES, TSESTree } from '@typescript-eslint/utils';
+import * as util from '../util';
+
+function isArrowFunctionTypeParameter(node: TSESTree.TSTypeParameter): boolean {
+  return (
+    node.parent?.type === AST_NODE_TYPES.TSTypeParameterDeclaration &&
+    node.parent.parent?.type === AST_NODE_TYPES.ArrowFunctionExpression
```
Comment (JoshuaKGoldberg) [inline packages/eslint-plugin/src/rules/no-redundant-generic-modifiers.ts:7]: For these `?.`s, the AST types says `node.parent` might not be defined but I don't know of a way to make that happen. Is this an incorrectness in the AST types?

**Review by bradzacher — COMMENTED** [review]
is some of this covered by [`no-unnecessary-type-constraint`](https://typescript-eslint.io/rules/no-unnecessary-type-constraint)?

**Comment by JoshuaKGoldberg** [comment]
Oh, dangit.  I knew this logic felt right and familiar -- and had no recollection of working on it a year and a half ago, apparently!

Still, no-unnecessary-type-constraint only deals with half of the linked issue. It doesn't tackle the _default_ modifiers -- just _contraint_ modifiers. I'll mark it as `deprecated` & `replacedBy` in this PR.

File: packages/eslint-plugin/src/rules/no-redundant-generic-modifiers.ts (hunk 1 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/src/rules/no-redundant-generic-modifiers.ts:7]: Ah, I'll just make this use the same style of ESQuery queries as the old rule.

**Comment by JoshuaKGoldberg** [comment]
Ah, actually, our own source code showed why this rule is fine as-is:

```ts
type ObjectLike<T = unknown> = Record<string, T>;

/**
 * Check if the variable contains an object strictly rejecting arrays
 * @param obj an object
 * @returns `true` if obj is an object
 */
function isObjectNotArray<T extends ObjectLike>(
  obj: unknown | unknown[],
): obj is T {
  return typeof obj === 'object' && !Array.isArray(obj);
}
```

Providing an explicit `= unknown` means the generic construct -in this case, `ObjectLike` can be used without an explicit type argument -in this case, `<T>`.

**Pull request closed without merging**